overlord/devicestate/handlers_install.go

```diff
@@ -1214,14 +1250,124 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
-	logger.Debugf("install-finish for %q on %v", systemLabel, onVolumes)
-	// TODO: use the seed to get gadget/kernel
-	// - install missing volumes structure content
-	// - install gadget assets
-	// - install kenrel
-	// - make system bootable (include writing modeenv)
+	logger.Debugf("starting install-finish for %q systemd on %v", systemLabel, onVolumes)
+
+	essentialTypes := []snap.Type{snap.TypeKernel, snap.TypeBase, snap.TypeGadget}
+	st.Unlock()
+	sys, snapInfos, snapSeeds, err := m.loadSystemAndEssentialSnaps(systemLabel, essentialTypes)
+	st.Lock()
+	if err != nil {
+		return err
+	}
+	// Mount gadget and kernel
+	mntPtForType := make(map[snap.Type]string)
+	for _, seedSn := range []*seed.Snap{snapSeeds[snap.TypeGadget], snapSeeds[snap.TypeKernel]} {
+		mntPt, rest, err := mountSeedSnap(seedSn)
+		mntPtForType[seedSn.EssentialType] = mntPt
+		if err != nil {
+			return err
+		}
+		defer func() {
+			errRest := rest()
+			if err == nil {
+				err = errRest
+			}
+		}()
+	}
+
+	// Unset revision means the snap will be local, we do something
+	// similar in the seedwriter.
+	// XXX: is this the right place?
+	for _, typ := range essentialTypes {
+		snInfo := snapInfos[typ]
+		if snInfo.Revision.Unset() {
+			snInfo.Revision = snap.R(-1)
+		}
+	}
+
+	// TODO useEncryption always false until setup encryption is implemented
+	useEncryption := false
+	installObserver, trustedInstallObserver, err := buildInstallObserver(sys.Model, mntPtForType[snap.TypeGadget], useEncryption)
+	if err != nil {
+		return err
+	}
 
-	return fmt.Errorf("finish install step not implemented yet")
+	logger.Debugf("writing content to partitions")
+	timings.Run(perfTimings, "install-finish", "Writing content to partitions", func(tm timings.Measurer) {
+		st.Unlock()
+		defer st.Lock()
+		_, err = installWriteContent(onVolumes, installObserver,
+			mntPtForType[snap.TypeGadget], mntPtForType[snap.TypeKernel], sys.Model, perfTimings)
+	})
+	if err != nil {
+		return fmt.Errorf("cannot write content: %v", err)
+	}
+
+	// Mount the partitions and find ESP partition
+	espMntDir := ""
+	numEsp := 0
+	for _, vol := range onVolumes {
+		for _, part := range vol.Structure {
+			if part.Filesystem == "" {
+				continue
+			}
+			mntPt := filepath.Join("/run/mnt", part.Name)
+			if err := syscall.Mount(part.Device, mntPt, part.Filesystem, 0, ""); err != nil {
+				return fmt.Errorf("cannot mount %q at %q: %v", part.Device, mntPt, err)
+			}
+			defer func() {
+				errUnmount := syscall.Unmount(mntPt, 0)
+				if errUnmount != nil {
+					logger.Noticef("cannot unmount %q: %v", mntPt, errUnmount)
+				}
+				// Do not clobber previous errors
+				if err == nil {
+					err = errUnmount
+				}
+			}()
+			if strings.Contains(strings.ToUpper(part.Type), "C12A7328-F81F-11D2-BA4B-00A0C93EC93B") {
+				espMntDir = mntPt
+				numEsp++
+			}
+		}
+	}
+	if numEsp != 1 {
+		return fmt.Errorf("there are %d ESP partitions, expected one", numEsp)
```

Comment:
this feels related/parallel to this code in gadget/install:  https://github.com/snapcore/snapd/blob/master/gadget/install/install.go#L353

it should be some form of helper there, is a bit strange to see syscall.Mount in devicestate